### PR TITLE
Convert chapter 10 (affine transformations)

### PR DESCRIPTION
**Pull request by gwilymk**
Converts the affine chapter to markdown. 

Closes https://github.com/gbadev-org/tonc/issues/22

### CODE REVIEW

**Comment by avivace** [comment]
Amazing, thanks @gwilymk . I closed the issue on the equations, you can go ahead with MathML ;)

**Comment by gwilymk** [comment]
mathml was a bit of a pain :upside_down_face:

**Comment by exelotl** [comment]
Thanks, this is looking really good so far!

Some notes from comparing side-by-side (md on the left, original (with dark theme) on the right):



* Subfigures can be done by assigning an ID to the containing tag for the whole figure rather than subfigures, and manually appending the letter e.g. `{@fig:foo}a` might render as _fig 1.2a_ .
* The two subfigures which shouldn't be side-by-side aren't, and the giant **A→**  isn't rendered right. Might be best to lift more of the original HTML to fix this?
* Vertical spacing around the equations may be needed (I guess you can just use `<br>` unless you feel like adding a CSS rule for equation blocks?)
* Letters in the formulae that used to be bold aren't anymore. I'm no expert mathematician but I guess this sort of thing is important to preserve?
* `{@fig:rotscale_good}` should be `{@fig:rotatescale_good}`.
* There's a couple of instances where the figure/table name ought to be capitalised but isn't (like near the bottom of the screenshot)

**Comment by gwilymk** [comment]
Sorry, was on holiday for the past two weeks. Plan is to finish this off tomorrow morning :)

**Comment by avivace** [comment]
@gwilymk amazing, thanks! Will take a look now

**Review by avivace — APPROVED** [review]
Looks great to me! The only thing I'm noticing is the different numbering on the figures: the first side by side gets 10.2 instead of 10.1. Is this intentional?

**Comment by gwilymk** [comment]
> the first side by side gets 10.2 instead of 10.1. Is this intentional?

It wasn't, but I couldn't work out why it was starting at 10.2 :/. It is the first figure since the one before is an equation. Tried a bunch of different things but none of them worked so hoped someone who knows how to use the referencing thing a bit better would have more luck :)

File: content/pages/affine.md [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+        </mtable>
+        <mo>]</mo>
+    </mrow>
+</math>
+
+Now, this is indeed a rotation and scale matrix. Unfortunately, it's also the <span class="rem">wrong one</span>! Or at least, it probably does not do what you'd expect. For example, consider the case with a scaling of <math><msub><mi>s</mi><mi>x</mi></msub></math> = 1.5, <math><msub><mi>s</mi><mi>y</mi></msub></math> = 1.0 and a rotation of α= 45. You'd probably expect something like {@fig:rotatescale}a, but what you'd actually get is {@fig:rotatescale}b. The sprite has rotated, but in the wrong direction, it has shrunk rather than expanded and there's an extra shear as well. Of course, you can always say that you meant for this to happen, but that's probably not quite true.
+
+<div style="display: flex; margin: 20px" markdown>
+<div class="cpt" style="width:160px;" markdown>
+![](img/affine/metr_rs_good.png){#fig:rotatescale}
```
Comment (LunarLambda) [inline content/pages/affine.md:58]: You're using the `#fig:rotatescale` id twice, resulting in the figure numbering being wrong. See [exelotl's comments](https://github.com/gbadev-org/tonc/pull/15#issuecomment-1712368738) for how to deal with subfigures.

Comment (gwilymk) [reply content/pages/affine.md:58]: Only mentioning the figure once fixed it here. Weird behaviour...

File: content/pages/affine.md [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+<div style="display: flex; align-items: center" markdown>
+<div class="cpt" style="width:128px;" markdown>
+![A metroid texture...](img/affine/metr_tex.png){#fig:metroid_texture}a
+**{*@fig:metroid_texture}a**: a texture.
+</div>
+
+<span style="font-size: 3em; display: inline-flex; flex-direction: column; padding: 10px" markdown>**A** →</span>
+
+<div class="cpt" style="width:128px;" markdown>
+![ ... mapped](img/affine/metr_texmapA.png){#fig:metroid_texture}b
```
Comment (LunarLambda) [inline content/pages/affine.md:125]: There is another duplicated figure id here.

File: content/pages/affine.md [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+![](img/affine/metr_rs_good.png){#fig:rotatescale}
+**{*@fig:rotatescale}a**: when you say ‘rotate and scale’, you probably expect this…
+</div>
+<div class="cpt" style="width:160px" markdown>
+![](img/affine/metr_rs_bad.png)
+**{*@fig:rotatescale}b**: but with **P** from {@eq:incorrect_transform_matrix}, this is what you get.
+</div>
+</div>
+
+Unfortunately, there is a lot of incorrect or misleading information on the transformation matrix around; the matrix of eq 10.1 is just one aspect of it. This actually starts with the moniker “Rot/Scale”, which does not fit with what actually occurs, continues with the fact that the terms used are never properly defined and that most people often just copy-paste from others without even considering checking whether the information is correct or not. The irony is that the principle reference document, GBATek, gives the correct descriptions of each of the elements, but somehow it got lost in the translation to matrix form in the tutorials.
```
Comment (LunarLambda) [inline content/pages/affine.md:67]: This needs to have {@eq:incorrect_transform_matrix}

File: content/pages/affine.md [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+    <mi>𝗾</mi>
+    <mo>=</mo>
+    <mi>𝗔</mi>
+    <mo>&middot;</mo>
+    <mi>𝗽</mi>
+</math>
+
+So how do you find **A**? Well, that's actually not that hard. The matrix is formed by lining up the transformed base vectors, which are **u** and **v** (this works in any number of dimensions, btw), so that gives us:
+
+<math id="eq:correct_transform_matrix" class="block">
```
Comment (LunarLambda) [inline content/pages/affine.md:97]: This should not have an `eq` id as it's never referred to in TONC's text. This is what's causing `{@eq:inverse_transform}` to show as 10.4 not 10.3

File: content/pages/affine.md [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+            <td>void <i>foo</i>_sheary(FOO_AFFINE *aff, FIXED hy);</td>
+            <td>Shear left-side down by <math><msub><mi>h</mi><mi>y</mi></msub></math></td>
+        </tr>
+    </tbody>
+</table>
+**{*@tbl:affine_functions}**: affine functions
+
+### Sample rot/scale function
+
+My code for a object version of the scale-then-rotate function (à la {@eq:transformation_matrix}) is given below. Note that it is from the computer's point of view, so that `sx` and `sy` scale down. Also, the alpha `alpha` uses 10000h/circle (i.e., the unit of α is π/8000h = 0.096 mrad, or 180/8000h = 0.0055°) and the sine lut is in .12f format, which is why the shifts by 12 are required. The background version is identical, except in name and type. If this were C++, templates would have been mighty useful here.
```
Comment (LunarLambda) [inline content/pages/affine.md:696]: This reference is broken. There's no equation with id `transformation_matrix`. You probably wanted `correct_matrix`.

**Review by LunarLambda — CHANGES_REQUESTED** [review]

**Comment by LunarLambda** [comment]
Filenames (i.e. things that were wrapped in `<tt>` tags in the original HTML) should be italicized rather than be wrapped in backticks.

Also, the math spacing is terrible on my system, but I'm not sure if this a font issue:

image

The figure/equation/table numbering looks all good now.

File: content/pages/affine.md [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+
+### Resolution
+
+I hope you spotted the crucial difference between the two points of view. **A** maps *from* texture space *to* screen space, while **B** does the exact opposite (i.e., <math><mi>𝗕</mi><mo>=</mo><msup><mi>𝗔</mi><mn>-1</mn></msup></math>). I think you know which one you should give the GBA by now. That's right: **P = B**, not **A**. This one bit of information is the crucial piece of the affine matrix puzzle.
+
+So now you can figure out **P**'s elements in two ways. You can stick to the human POV and invert the matrix at the end. That's why I gave you the inverses of the affine transformations as well. You could also try to see things in the GBA's way and get the right matrix directly. Tonc's main affine functions (`tonc_video.h`, `tonc_obj_affine.c` and `tonc_bg_affine.c`) do things the GBA way, setting **P** directly; but inverted functions are also available using an "`_inv`" affix. Mind you, these are a little slower. Except for when scaling is involved; then it's a *lot* slower.
```
Comment (avivace) [inline content/pages/affine.md:439]: Filenames here should go in italic

**Review by avivace — CHANGES_REQUESTED** [review]

**Pull request merged**